File: ts/src/production-traces/contract/branded-ids.ts
import { ulid } from "ulid";

declare const brand: unique symbol;
type Brand<T, B> = T & { readonly [brand]: B };

// Branded IDs introduced by the production-traces contract.
export type ProductionTraceId = Brand<string, "ProductionTraceId">;
export type AppId             = Brand<string, "AppId">;
export type UserIdHash        = Brand<string, "UserIdHash">;
export type SessionIdHash     = Brand<string, "SessionIdHash">;
export type FeedbackRefId     = Brand<string, "FeedbackRefId">;

// Crockford base32: 0-9 A-H J K M N P-T V-Z (excludes I L O U). ULID is 26 chars.
const ULID_RE = /^[0-9A-HJKMNP-TV-Z]{26}$/;
// AppId: lowercase alnum start + [a-z0-9_-]* — path-safe and grep-friendly.
const APP_ID_RE = /^[a-z0-9][a-z0-9_-]*$/;
// SHA-256 hex — 64 chars, lowercase.
const SHA256_HEX_RE = /^[0-9a-f]{64}$/;

export function newProductionTraceId(): ProductionTraceId {
  return ulid() as ProductionTraceId;
}

export function parseProductionTraceId(input: string): ProductionTraceId | null {
  return ULID_RE.test(input) ? (input as ProductionTraceId) : null;
}

export function parseAppId(input: string): AppId | null {
  if (input === ".." || input.includes("/") || input.includes("\\")) return null;
  return APP_ID_RE.test(input) ? (input as AppId) : null;
}

export function parseUserIdHash(input: string): UserIdHash | null {
  return SHA256_HEX_RE.test(input) ? (input as UserIdHash) : null;
}

export function parseSessionIdHash(input: string): SessionIdHash | null {
  return SHA256_HEX_RE.test(input) ? (input as SessionIdHash) : null;
}

export function parseFeedbackRefId(input: string): FeedbackRefId | null {
  // Opaque customer-supplied identifier: reject only if fully whitespace or empty.
  if (input.trim().length === 0) return null;
  return input as FeedbackRefId;
}

// Re-exports from control-plane/contract/ for ergonomic downstream imports.
// Do NOT duplicate — downstream consumers should import these brands from here
// so production-traces stays the single import origin for this contract.
export {
  parseEnvironmentTag,
  defaultEnvironmentTag,
  parseContentHash,
  parseScenario,
} from "../../control-plane/contract/branded-ids.js";
export type {
  EnvironmentTag,
  ContentHash,
  Scenario,
} from "../../control-plane/contract/branded-ids.js";

File: autocontext/src/autocontext/production_traces/contract/branded_ids.py
"""Branded / constrained aliases for production-trace IDs.

Pydantic v2 picks up the ``pattern`` constraint on parse, giving us schema-level
validation equivalent to the TS branded-id parsers. ``NewType`` is used for the
opaque ``FeedbackRefId`` since it has no pattern constraint.
"""

from __future__ import annotations

from typing import Annotated, NewType

from pydantic import StringConstraints

# ULID pattern — Crockford base32 excludes I/L/O/U. 26 chars, uppercase.
ProductionTraceId = Annotated[str, StringConstraints(pattern=r"^[0-9A-HJKMNP-TV-Z]{26}$")]

# AppId: slug-like, lowercase, non-empty, path-safe.
AppId = Annotated[str, StringConstraints(pattern=r"^[a-z0-9][a-z0-9_-]*$")]

# SHA-256 hex (lowercase, exactly 64 chars).
UserIdHash = Annotated[str, StringConstraints(pattern=r"^[0-9a-f]{64}$")]
SessionIdHash = Annotated[str, StringConstraints(pattern=r"^[0-9a-f]{64}$")]

# Opaque customer-supplied reference — no format enforced beyond non-emptiness.
FeedbackRefId = NewType("FeedbackRefId", str)

# Reused from shared defs / control-plane conventions.
EnvironmentTag = Annotated[str, StringConstraints(pattern=r"^[a-zA-Z0-9][a-zA-Z0-9_-]*$")]
ContentHash = Annotated[str, StringConstraints(pattern=r"^sha256:[0-9a-f]{64}$")]
Scenario = Annotated[str, StringConstraints(pattern=r"^[a-z0-9][a-z0-9_-]*$")]

__all__ = [
    "AppId",
    "ContentHash",
    "EnvironmentTag",
    "FeedbackRefId",
    "ProductionTraceId",
    "Scenario",
    "SessionIdHash",
    "UserIdHash",
]
